src/main/java/com/minecolonies/coremod/entity/ai/citizen/guard/AbstractEntityAIGuard.java

```diff
@@ -1,654 +1,650 @@
-
-    /**
-     * Last direction the guard headed to.
-     */
-    private Tuple<EnumFacing, EnumFacing> lastDirection = null;
-
-    /**
-     * Sets up some important skeleton stuff for every ai.
+     * Creates the abstract part of the AI.
+     * Always use this constructor!
      *
-     * @param job the job class
+     * @param job the job to fulfill
      */
-    protected AbstractEntityAIGuard(@NotNull final JobGuard job)
+    public AbstractEntityAIGuard(@NotNull final J job)
     {
         super(job);
         super.registerTargets(
-          new AITarget(this::checkIfExecute, this::getState),
-          new AITarget(IDLE, () -> START_WORKING),
-          new AITarget(START_WORKING, () -> GUARD_RESTOCK),
-          new AITarget(GUARD_GATHERING, this::gathering)
+          new AITarget(IDLE, START_WORKING),
+          new AITarget(START_WORKING, this::startWorkingAtOwnBuilding),
+          new AITarget(PREPARING, this::prepare),
+          new AITarget(DECIDE, this::decide),
+          new AITarget(GUARD_ATTACK_PROTECT, this::attackProtect),
+          new AITarget(GUARD_ATTACK_PHYSICAL, this::attackPhyisical),
+          new AITarget(GUARD_ATTACK_RANGED, this::attackRanged)
         );
+        worker.setCanPickUpLoot(true);
     }
 
     /**
-     * Checks if the worker is in a state where he can execute well.
-     *
-     * @return true if so.
+     * Can be overridden in implementations.
+     * <p>
+     * Here the AI can check if the armour have to be re rendered and do it.
      */
-    private boolean checkIfExecute()
+    @Override
+    protected void updateRenderMetaData()
     {
-        final AbstractBuilding building = getOwnBuilding();
-        if (!(building instanceof AbstractBuildingGuards))
-        {
-            return true;
-        }
+        updateArmor();
+    }
 
-        if (!((AbstractBuildingGuards) building).shallRetrieveOnLowHealth())
-        {
-            return false;
-        }
+    protected abstract int getAttackRange();
 
-        if (worker.getHealth() > LOW_HEALTH)
+    /**
+     * Redirects the herder to their building.
+     *
+     * @return The next {@link AIState}.
+     */
+    private AIState startWorkingAtOwnBuilding()
+    {
+        worker.setLatestStatus(new TextComponentTranslation(TranslationConstants.COM_MINECOLONIES_COREMOD_STATUS_WORKER_GOINGTOHUT));
+        if (walkToBuilding())
         {
-            return false;
+            return getState();
         }
-
-        worker.isWorkerAtSiteWithMove(building.getLocation(), PATH_CLOSE);
-        return true;
+        return PREPARING;
     }
 
     /**
-     * Goes back to the building and tries to take armour from it when he hasn't in his inventory.
+     * Prepares the herder for herding
      *
-     * @return the next state to go to.
+     * @return The next {@link AIState}.
      */
-    protected AIState goToBuilding()
+    private AIState prepare()
     {
-        if (walkToBuilding())
+        setDelay(20 * 5);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Assign this magic number 20 to a well-named constant, and use the constant instead. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Assign this magic number 5 to a well-named constant, and use the constant instead. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)